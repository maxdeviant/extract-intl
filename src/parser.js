import * as parser from '@babel/parser'
import traverse from '@babel/traverse'

const MESSAGE_TAGS = ['FormattedMessage', 'FormattedHTMLMessage']

function isJSXElement(node) {
  return node.type === 'JSXElement'
}

function isFormattedMessage(node) {
  return (
    isJSXElement(node) && MESSAGE_TAGS.includes(node.openingElement.name.name)
  )
}

function makeFindPropByName(propName) {
  return function findPropByName(jsxElementNode) {
    return jsxElementNode.openingElement.attributes.find(
      attr => attr.name && attr.name.name === propName
    )
  }
}

function extractStringValue(propNode) {
  if (propNode.value.type === 'JSXExpressionContainer') {
    const expressionNode = propNode.value
    if (expressionNode.expression.type === 'TemplateLiteral') {
      return expressionNode.expression.quasis[0].value.raw
    }

    return expressionNode.expression.value
  }

  return propNode.value.value
}

function findId(formattedMessageNode) {
  const idNode = makeFindPropByName('id')(formattedMessageNode)
  if (!idNode) {
    return void 0
  }

  return extractStringValue(idNode)
}

function findDefaultMessage(formattedMessageNode) {
  const defaultMessageNode = makeFindPropByName('defaultMessage')(
    formattedMessageNode
  )
  if (!defaultMessageNode) {
    return void 0
  }

  return extractStringValue(defaultMessageNode)
}

function findDescription(formattedMessageNode) {
  const descriptionNode = makeFindPropByName('description')(
    formattedMessageNode
  )
  if (!descriptionNode) {
    return void 0
  }

  return extractStringValue(descriptionNode)
}

export const DEFAULT_PARSER_OPTIONS = {
  sourceType: 'module',
  plugins: ['jsx']
}

/**
 * Returns an object containing all of the translatable messages in the file contents.
 *
 * @param contents The file contents to search.
 */
export function parse(contents, parserOptions) {
  if (!parserOptions) {
    parserOptions = DEFAULT_PARSER_OPTIONS
  }

  const ast = parser.parse(contents, parserOptions)

  const messageDescriptors = []

  traverse(ast, {
    enter(path) {
      const { node } = path
      if (isFormattedMessage(node)) {
        messageDescriptors.push({
          id: findId(node) || '',
          defaultMessage: findDefaultMessage(node) || '',
          values: {},
          description: findDescription(node) || ''
        })
      }
    }
  })

  const messages = {}
  const duplicates = []
  messageDescriptors
    .filter(messageDescriptor => !!messageDescriptor.id)
    .forEach(messageDescriptor => {
      const { id: messageId, ...message } = messageDescriptor
      if (messageId in messages) {
        duplicates.push(messageId)
      }
      messages[messageId] = message
    })

  return {
    messages,
    duplicates
  }
}
