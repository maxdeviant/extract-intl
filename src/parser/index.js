import md5 from 'md5'
import { basename, extname } from 'path'
import { parseComponents } from './component'
import { parseFunctions } from './function'

/**
 * Returns an object containing all of the translatable messages in the file contents.
 *
 * @param filepath The path to the file.
 * @param contents The file contents to search.
 */
export function parse(filepath, contents) {
  const filehash = md5(filepath).slice(0, 7)
  const filename = basename(filepath, extname(filepath))
  const messageDescriptors = [
    ...parseComponents(contents),
    ...parseFunctions(contents)
  ]

  const messages = {}
  const duplicates = []
  messageDescriptors.forEach((messageDescriptor, index) => {
    const messageId = `${filehash}.${filename}.${index + 1}`
    if (messageId in messages) {
      duplicates.push(messageId)
    }
    messages[messageId] = messageDescriptor
  })

  return {
    messages,
    duplicates
  }
}
