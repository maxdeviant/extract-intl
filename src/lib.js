import fs, { read } from 'fs'
import { parse } from './parser'

export function extractIntl(files) {
  return new Promise((resolve, reject) => {
    if (!Array.isArray(files)) {
      files = [files]
    }

    return Promise.all(
      files.map(filepath =>
        readFile(filepath).then(contents => parse(filepath, contents))
      )
    )
      .then(files => {
        return resolve(
          files.reduce(
            (accumulator, { messages, duplicates }) => ({
              messages: {
                ...accumulator.messages,
                ...messages
              },
              duplicates: [...accumulator.duplicates, ...duplicates]
            }),
            {
              messages: {},
              duplicates: []
            }
          )
        )
      })
      .catch(err => {
        return reject(err)
      })
  })
}

function readFile(filePath) {
  return new Promise((resolve, reject) => {
    fs.readFile(filePath, 'utf-8', (err, data) => {
      if (err) {
        return reject(err)
      }

      return resolve(data)
    })
  })
}
