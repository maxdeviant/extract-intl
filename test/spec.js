import path from 'path'
import { extractIntl } from '../src/lib'

describe('extract-intl', () => {
  describe('FormattedMessage', () => {
    it('should extract a message', () => {
      const fixturesPath = path.join(
        __dirname,
        '/fixtures/formatted-message/simple.jsx'
      )

      return expect(extractIntl(fixturesPath)).resolves.toMatchSnapshot()
    })

    it('should extract values', () => {
      const fixturesPath = path.join(
        __dirname,
        '/fixtures/formatted-message/values.jsx'
      )

      return expect(extractIntl(fixturesPath)).resolves.toMatchSnapshot()
    })

    it('should extract multiple messages', () => {
      const fixturesPath = path.join(
        __dirname,
        '/fixtures/formatted-message/multiple.jsx'
      )

      return expect(extractIntl(fixturesPath)).resolves.toMatchSnapshot()
    })

    it('should extract from multiple files', () => {
      const fixturesPaths = [
        path.join(__dirname, '/fixtures/formatted-message/a.jsx'),
        path.join(__dirname, '/fixtures/formatted-message/b.jsx')
      ]

      return expect(extractIntl(fixturesPaths)).resolves.toMatchSnapshot()
    })

    it('should handle other string literal formats', () => {
      const fixturesPath = path.join(
        __dirname,
        '/fixtures/formatted-message/string-literals.jsx'
      )

      return expect(extractIntl(fixturesPath)).resolves.toMatchSnapshot()
    })
  })

  describe('formatMessage', () => {
    it('should extract a message', () => {
      const fixturesPath = path.join(
        __dirname,
        '/fixtures/format-message/simple.jsx'
      )

      return expect(extractIntl(fixturesPath)).resolves.toMatchSnapshot()
    })

    it('should extract values', () => {
      const fixturesPath = path.join(
        __dirname,
        '/fixtures/format-message/values.jsx'
      )
      return expect(extractIntl(fixturesPath)).resolves.toMatchSnapshot()
    })

    it('should extract multiple messages', () => {
      const fixturesPath = path.join(
        __dirname,
        '/fixtures/format-message/multiple.jsx'
      )
      return expect(extractIntl(fixturesPath)).resolves.toMatchSnapshot()
    })

    it('should extract from multiple files', () => {
      const fixturesPaths = [
        path.join(__dirname, '/fixtures/format-message/a.jsx'),
        path.join(__dirname, '/fixtures/format-message/b.jsx')
      ]

      return expect(extractIntl(fixturesPaths)).resolves.toMatchSnapshot()
    })

    it('should handle properties in varying orders', () => {
      const fixturesPath = path.join(
        __dirname,
        '/fixtures/format-message/order.jsx'
      )

      return expect(extractIntl(fixturesPath)).resolves.toMatchSnapshot()
    })

    it('should handle other string literal formats', () => {
      const fixturesPath = path.join(
        __dirname,
        '/fixtures/format-message/string-literals.jsx'
      )

      return expect(extractIntl(fixturesPath)).resolves.toMatchSnapshot()
    })

    it('should handle weird formatting', () => {
      const fixturesPath = path.join(
        __dirname,
        '/fixtures/format-message/formatting.jsx'
      )

      return expect(extractIntl(fixturesPath)).resolves.toMatchSnapshot()
    })
  })
})
