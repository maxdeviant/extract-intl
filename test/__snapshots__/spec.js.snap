// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extract-intl FormattedMessage should extract a message 1`] = `
Object {
  "duplicates": Array [],
  "messages": Object {
    "b36a868.simple.1": Object {
      "defaultMessage": "My List Item",
      "description": "",
      "values": Object {},
    },
  },
}
`;

exports[`extract-intl FormattedMessage should extract from multiple files 1`] = `
Object {
  "duplicates": Array [],
  "messages": Object {
    "3e94d25.a.1": Object {
      "defaultMessage": "Component A",
      "description": "",
      "values": Object {},
    },
    "3e94d25.a.2": Object {
      "defaultMessage": "Hello there!",
      "description": "",
      "values": Object {},
    },
    "7e0081b.b.1": Object {
      "defaultMessage": "Component B",
      "description": "",
      "values": Object {},
    },
  },
}
`;

exports[`extract-intl FormattedMessage should extract multiple messages 1`] = `
Object {
  "duplicates": Array [],
  "messages": Object {
    "4c77235.multiple.1": Object {
      "defaultMessage": "My List Item",
      "description": "",
      "values": Object {},
    },
    "4c77235.multiple.2": Object {
      "defaultMessage": "This is my body text.",
      "description": "",
      "values": Object {},
    },
  },
}
`;

exports[`extract-intl FormattedMessage should extract values 1`] = `
Object {
  "duplicates": Array [],
  "messages": Object {
    "0fde777.values.1": Object {
      "defaultMessage": "My {adjective} List Item",
      "description": "",
      "values": Object {},
    },
  },
}
`;

exports[`extract-intl FormattedMessage should handle other string literal formats 1`] = `
Object {
  "duplicates": Array [],
  "messages": Object {
    "4569682.string-literals.1": Object {
      "defaultMessage": "This footer is bangin'",
      "description": "",
      "values": Object {},
    },
    "4569682.string-literals.2": Object {
      "defaultMessage": "You have {messageCount, plural, one {# new message} other {# new messages}}!",
      "description": "",
      "values": Object {},
    },
    "4569682.string-literals.3": Object {
      "defaultMessage": "Copyright (c) 2016",
      "description": "",
      "values": Object {},
    },
  },
}
`;

exports[`extract-intl formatMessage should extract a message 1`] = `
Object {
  "duplicates": Array [],
  "messages": Object {
    "12e95ab.simple.1": Object {
      "defaultMessage": "Copyright (c) 2016",
      "description": "",
      "values": Object {},
    },
  },
}
`;

exports[`extract-intl formatMessage should extract from multiple files 1`] = `
Object {
  "duplicates": Array [],
  "messages": Object {
    "3455bcd.a.1": Object {
      "defaultMessage": "Component A",
      "description": "",
      "values": Object {},
    },
    "3455bcd.a.2": Object {
      "defaultMessage": "Hello there!",
      "description": "",
      "values": Object {},
    },
    "fbf0c92.b.1": Object {
      "defaultMessage": "Component B",
      "description": "",
      "values": Object {},
    },
  },
}
`;

exports[`extract-intl formatMessage should extract multiple messages 1`] = `
Object {
  "duplicates": Array [],
  "messages": Object {
    "3d8bbc8.multiple.1": Object {
      "defaultMessage": "Why does a footer need a title?",
      "description": "",
      "values": Object {},
    },
    "3d8bbc8.multiple.2": Object {
      "defaultMessage": "Copyright (c) 2016",
      "description": "",
      "values": Object {},
    },
  },
}
`;

exports[`extract-intl formatMessage should extract values 1`] = `
Object {
  "duplicates": Array [],
  "messages": Object {
    "fcfa8e7.values.1": Object {
      "defaultMessage": "Copyright (c) {year}",
      "description": "",
      "values": Object {},
    },
  },
}
`;

exports[`extract-intl formatMessage should handle other string literal formats 1`] = `
Object {
  "duplicates": Array [],
  "messages": Object {
    "be03946.string-literals.1": Object {
      "defaultMessage": "This footer is bangin'",
      "description": "",
      "values": Object {},
    },
    "be03946.string-literals.2": Object {
      "defaultMessage": "Copyright (c) 2016",
      "description": "",
      "values": Object {},
    },
    "be03946.string-literals.3": Object {
      "defaultMessage": "ALL RIGHTS RESERVED",
      "description": "",
      "values": Object {},
    },
  },
}
`;

exports[`extract-intl formatMessage should handle properties in varying orders 1`] = `
Object {
  "duplicates": Array [],
  "messages": Object {
    "48ffe03.order.1": Object {
      "defaultMessage": "UserProfile",
      "description": "The title for the user profile.",
      "values": Object {},
    },
  },
}
`;

exports[`extract-intl formatMessage should handle weird formatting 1`] = `
Object {
  "duplicates": Array [],
  "messages": Object {
    "8f52cd4.formatting.1": Object {
      "defaultMessage": "Copyright (c) 2016",
      "description": "",
      "values": Object {},
    },
    "8f52cd4.formatting.2": Object {
      "defaultMessage": "ALL RIGHTS RESERVED",
      "description": "",
      "values": Object {},
    },
  },
}
`;
